bonnie++.8
.TH bonnie++ 8
.SH "NAME"
bonnie++ \- program to test hard drive performance.

.SH "SYNOPSIS"
.B bonnie++
.I [\-d dir] [\-s size(MiB)[:chunk\-size(b)]]
.I [\-n number\-to\-stat(*1024)[:max\-size[:min\-size][:num\-directories]]]
.I [\-m machine\-name] [\-r ram\-size\-in\-MiB] [\-x number\-of\-tests]
.I [\-u uid\-to\-use:gid\-to\-use] [\-g gid\-to\-use]
.I [\-q] [\-f] [\-b] [\-D] [\-p processes | \-y]

.SH "DESCRIPTION"
This manual page documents briefly the
.BR bonnie++ ,
program.
.P
Bonnie++ is a program to test hard drives and file systems for performance or
the lack therof. There are a many different types of file system operations
which different applications use to different degrees. Bonnie++ tests some of
them and for each test gives a result of the amount of work done per second
and the percentage of CPU time this took. For performance results higher
numbers are better, for CPU usage lower are better (NB a configuration scoring
a performance result of 2000 and a CPU result of 90% is better in terms of CPU
use than a configuration delivering performance of 1000 and CPU usage of 60%).
.P
There are two sections to the program's operations. The first is to test the
IO throughput in a fashion that is designed to simulate some types of database
applications. The second is to test creation, reading, and deleting many small
files in a fashion similar to the usage patterns of programs such as Squid or
INN.
.P
All the details of the tests performed by Bonnie++ are contained in the file
.BR /usr/share/doc/bonnie++/readme.html

.SH "OPTIONS"
For Bonnie++ every option is of the form of a hyphen followed by a letter and
then the next parameter contains the value.
.TP
.B \-d
the directory to use for the tests.
.TP
.B \-s
the size of the file(s) for IO performance measures in megabytes. If the size
is greater than 1G then multiple files will be used to store the data, and
each file will be up to 1G in size.  This parameter may include the chunk size
seperated from the size by a colon.  The chunk\-size is measured in bytes and
must be a power of two from 256 to 1048576, the default is 8192.  NB You can
specify the size in giga\-bytes or the chunk\-size in kilo\-bytes if you add
\f3g\f1 or \f3k\f1 to the end of the number respectively.

If the specified size is 0 then this test will be skipped.
.TP
.B \-n
the number of files for the file creation test. This is measured in multiples
of 1024 files. This is because no\-one will want to test less than 1024 files,
and we need the extra space on braille displays.

The maximum value for this is 262143, that is 268,434,432 files - it should be
enough.

If the specified number is 0 then this test will be skipped.

The default for this test is to test with 0 byte files. To use files of other
sizes you can specify
.B number:max:min:num\-directories
where
.B max
is the maximum size and
.B min
is the minimum size (both default to 0 if not specified). If minimum and
maximum sizes are specified then every file will have a random size from the
range
.B min..max
inclusive.  If you specify a number of directories then the files will be
evenly distributed amoung that many sub\-directories.

If
.B max
is
.B \-1
then hard links will be created instead of files.
If
.B max
is
.B \-2
then soft links will be created instead of files.


.TP
.B \-m
name of the machine \- for display purposes only.

.TP
.B \-r
RAM size in megabytes. If you specify this the other parameters will be
checked to ensure they make sense for a machine of that much RAM. You should
not need to do this in general use as it should be able to discover the RAM
size. NB If you specify a size of 0 then all checks will be disabled...

.TP
.B \-x
number of test runs. This is useful if you want to perform more than one test.
It will dump output continuously in CSV format until either the number of
tests have been completed, or it is killed.

.TP
.B \-u
user\-id to use.  When running as root specify the UID to use for the tests.  It
is not recommended to use root (since the occasion when a Bonnie++ bug wiped
out someone's system), so if you really want to run as root then use
.B \-u root.
Also if you want to specify the group to run as then use the
.B user:group
format.  If you specify a user by name but no group then the primary group of
that user will be chosen.  If you specify a user by number and no group then
the group will be
.B nogroup.

.TP
.B \-g
group\-id to use.  Same as using
.B :group
for the
.B \-u
parameter, just a different way to specify it for compatibility with other
programs.

.TP
.B \-q
quiet mode. If specified then some of the extra informational messages will be
suppressed.

.TP
.B \-f
fast mode, skips per\-char IO tests.

.TP
.B \-b
no write buffering.  fsync() after every write.

.TP
.B \-p
number of processes to serve semaphores for.  This is used to create the
semaphores for synchronising multiple Bonnie++ processes.  All the processes
which are told to use the semaphore with
.B \-y
will start each test at the same time.  Use the value \-1 to delete the
semaphore.

.TP
.B \-y
wait for semaphore before each test.

.TP
.B \-D
use direct IO (O_DIRECT) for the bulk IO tests

.P

.SH "MULTIPLE PROCESSES"
Run the following commands to run three copies of Bonnie++ simultaneously:
.P
bonnie++ -p3
.P
bonnie++ -y > out1 &
.P
bonnie++ -y > out2 &
.P
bonnie++ -y > out3 &

.SH "OUTPUT"
The primary output is plain\-text in 80 columns which is designed to fit well
when pasted into email and which will work well with Braille displays.
.P
The second type of output is CSV (Comma Seperated Values). This can easily be
imported into any spread\-sheet or database program. Also I have included
the programs
.B bon_csv2html
and
.B bon_csv2txt
to convert CSV data to HTML and plain\-ascii respectively.
.P
For every test two numbers are reported, the amount of work done (higher
numbers are better) and the percentage of CPU time taken to perform the work
(lower numbers are better). If a test completes in less than 500ms then
the output will be displayed as "++++". This is because such a test result
can't be calculated accurately due to rounding errors and I would rather
display no result than a wrong result.

.SH "AUTHOR"
This program, its manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>, parts of the program are based on the
work of Tim Bray <tbray@textuality.com>.
.P
The documentation, the Perl scripts, and all the code for testing the creation
of thousands of files was written by Russell Coker, but the entire package is
under joint copyright with Tim Bray.

.SH "SIGNALS"
Handles SIGINT and does a cleanup (which may take some time), a second SIGINT
or a SIGQUIT will cause it to immediately die.
.P
SIGXCPU and SIGXFSZ act like SIGINT.
.P
Ignores SIGHUP.


.SH "BUGS"
The random file sizes will add up to different values for different test runs.
I plan to add some code that checks the sum and ensures that the sum of the
values will be the same on seperate runs.

.SH "AVAILABILITY"
The source is available from http://www.coker.com.au/bonnie++ .
.P
See http://etbe.coker.com.au/category/benchmark for further information.

.SH "SEE ALSO"
.BR bon_csv2html (1),
.BR bon_csv2txt (1)


debian/changelog
bonnie++ (1.04) UNRELEASED; urgency=medium

  * Allow 10000 files for IO tests (used to be 1000).
  * Use debhelper compat level 11.

 -- Russell Coker <russell@coker.com.au>  Tue, 05 Sep 2017 14:27:02 +1000

bonnie++ (1.03f) unstable; urgency=low

  * Fix a minor bug in setting the GID.
  * Make it build on Solaris by checking for O_DIRECT being defined.
  * The option -n9765 or greater caused the small file test to abort,
    to increase this while maintaining result compatability (same file length)
    I changed the format to hexadecimal, the limit is now 262,143.  I also
    made it handle errors better.  Thanks Nick Dokos of HP.

 -- Russell Coker <russell@coker.com.au>  Fri, 03 Jul 2009 11:58:38 +1000

bonnie++ (1.03e) unstable; urgency=low

  * Added the -D option to bonnie++ to use direct IO (O_DIRECT) for bulk IO.
    Thanks to Dave Murch from Violin Memory - http://violin-memory.com/

 -- Russell Coker <russell@coker.com.au>  Wed, 10 Dec 2008 09:08:00 +1100

bonnie++ (1.03d) unstable; urgency=low

  * Made it compile with GCC 4.3.
  * Added the URL to the man pages.
  * Made zcav display the total time and average speed for a loop and also the
    time that each loop completed.
  * Made the zcav default block size 256M to cope with larger and faster disks.
  * Made the zcav output units be in MiB/s and GiB.
  * Improved the Bonnie++ man page in regard to the -y and -p options.
  * Use DH_COMPAT=4 and dh_installman.

 -- Russell Coker <russell@coker.com.au>  Wed, 23 Jul 2008 10:54:24 +1000

bonnie++ (1.03c) unstable; urgency=low

  * Fixed some failures to check the return code of chdir().

 -- Russell Coker <russell@coker.com.au>  Sun, 27 Jan 2008 15:31:00 +1100

bonnie++ (1.03b.1) unstable; urgency=low

  * Fixed typos in man page.
    Closes: #133714
    Closes: #449596
  * Documented why UID==0 is bad.
    Closes: #182023

 -- Russell Coker <russell@coker.com.au>  Wed,  5 Dec 2007 09:09:00 +1100

bonnie++ (1.03b) unstable; urgency=low

  * Added support for testing write speed to zcav.

  * Documented the default chunk size in bonnie++.8.

  * Improved error handling for chunk size.

 -- Russell Coker <russell@coker.com.au>  Mon,  3 Dec 2007 10:13:00 +1100

bonnie++ (1.03a) unstable; urgency=low

  * Fixed the bugs in GCC 3.2 compilation and compiled with GCC 3.2.  Results
    will be slightly lower now, but it's still a fair test of compiler/OS...

 -- Russell Coker <russell@coker.com.au>  Fri, 10 Jan 2002 15:20:00 +0100

bonnie++ (1.03) unstable; urgency=low

  * Fixed a bug in error handling during a stat test with multiple directories.

 -- Russell Coker <russell@coker.com.au>  Fri,  6 Dec 2002 12:52:00 +0100

bonnie++ (1.02d) unstable; urgency=low

  * Fixed a bug where it would SEGV if you specified more than 999 directories
    (now it handles 99,999 and displays an error message if you ask for more).
    Closes: #171507

  * Made it build on Irix.

  * Fixed <vector> and <algo> checks to work with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Tue,  3 Dec 2002 15:12:00 +0100

bonnie++ (1.02c) unstable; urgency=low

  * Improved the RPM packaging.

  * Added support for detecting memory size on AIX.

  * I changed minimum time for a test to be considered valid to half a second
    but forgot to change the man page, changed it now.

  * Fixed an error in calculating seeks, it used to add an extra 5 seconds to
    the test time before calculation and therefore under-reported the results.

 -- Russell Coker <russell@coker.com.au>  Mon, 22 Apr 2002 00:33:00 +0200

bonnie++ (1.02b) unstable; urgency=low

  * Made the signal handler extern "C".

  * Fixed a spelling error in the description.
    Closes: #124462

 -- Russell Coker <russell@coker.com.au>  Wed, 19 Dec 2001 07:55:00 +0100

bonnie++ (1.02a) unstable; urgency=low

  * Removed optind=0, which caused problems on BSD.

 -- Russell Coker <russell@coker.com.au>  Tue,  6 Nov 2001 17:23:00 +0100

bonnie++ (1.02-1) unstable; urgency=low

  * Updated to latest debian standards version and follow the latest Debian
    policy on stripping and debug generation.

 -- Russell Coker <russell@coker.com.au>  Sat,  6 Oct 2001 19:06:39 +0200

bonnie++ (1.02) unstable; urgency=low

  * Fixed the cause of a minor warning on AIX.

 -- Russell Coker <russell@coker.com.au>  Wed, 19 Sep 2001 17:05:01 +0200

bonnie++ (1.01d) unstable; urgency=low

  * Fixed a bug where bonnie++ would SEGV if ^C was pressed at the start of the
    file creation tests.

  * Clarified the -r option to bonnie++ in the man page.

  * Minor fix to command line checking for chunk size.  Now works for 1M chunks.

 -- Russell Coker <russell@coker.com.au>  Thu, 26 Jul 2001 21:46:18 +0200

bonnie++ (1.01c) unstable; urgency=low

  * Made the -d option to bonnie++ work with relative paths.  Now it does
    chdir() to the specifed directory and proceeds as usual.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 May 2001 14:36:17 +0100

bonnie++ (1.01b-1) unstable; urgency=low

  * Compiled with GCC 2.95 for Debian.
    Closes:#94349


bonnie++.cpp

/*
 * COPYRIGHT NOTICE:
 * Copyright (c) Tim Bray, 1990.
 * Copyright (c) Russell Coker, 1999.  I have updated the program, added
 * support for >2G on 32bit machines, and tests for file creation.
 * Licensed under the GPL version 2.0.
 * DISCLAIMER:
 * This program is provided AS IS with no warranty of any kind, and
 * The author makes no representation with respect to the adequacy of this
 *  program for any particular purpose or with respect to its adequacy to
 *  produce any particular result, and
 * The author shall not be liable for loss or damage arising out of
 *  the use of this program regardless of how sustained, and
 * In no event shall the author be liable for special, direct, indirect
 *  or consequential damage, loss, costs or fees or expenses of any
 *  nature or kind.
 */

#ifdef OS2
#define INCL_DOSFILEMGR
#define INCL_DOSMISC
#define INCL_DOSQUEUES
#define INCL_DOSPROCESS
#include <os2.h>
#else
#include <sys/wait.h>
#include <unistd.h>
#endif
#include <sys/time.h>
#include <time.h>
#include <stdlib.h>
#include "bonnie.h"
#include "bon_io.h"
#include "bon_file.h"
#include "bon_time.h"
#include "semaphore.h"
#include <pwd.h>
#include <grp.h>
#include <ctype.h>
#include <string.h>
#include <sys/utsname.h>
#include <signal.h>

#ifdef AIX_MEM_SIZE
#include <cf.h>
#include <sys/cfgodm.h>
#include <sys/cfgdb.h>
#endif

void usage();

class CGlobalItems
{
public:
  bool quiet;
  bool fast;
  bool sync_bonnie;
#ifdef O_DIRECT
  bool use_direct_io;
#endif
  BonTimer timer;
  int ram;
  Semaphore sem;
  char *name;
  bool bufSync;
  int  chunk_bits;
  int chunk_size() const { return m_chunk_size; }
  bool *doExit;
  void set_chunk_size(int size)
    { delete m_buf; pa_new(size, m_buf, m_buf_pa); m_chunk_size = size; }

  // Return the page-aligned version of the local buffer
  char *buf() { return m_buf_pa; }

  CGlobalItems(bool *exitFlag);
  ~CGlobalItems() { delete name; delete m_buf; }

  void decrement_and_wait(int nr_sem);

  void SetName(CPCCHAR path)
  {
    delete name;
    name = new char[strlen(path) + 15];
#ifdef OS2
    ULONG myPid = 0;
    DosQuerySysInfo(QSV_FOREGROUND_PROCESS, QSV_FOREGROUND_PROCESS
                  , &myPid, sizeof(myPid));
#else
    pid_t myPid = getpid();
#endif
    sprintf(name, "%s/Bonnie.%d", path, int(myPid));
  }
private:
  int m_chunk_size;
  char *m_buf;     // Pointer to the entire buffer
  char *m_buf_pa;  // Pointer to the page-aligned version of the same buffer

  CGlobalItems(const CGlobalItems &f);
  CGlobalItems & operator =(const CGlobalItems &f);

  // Implement a page-aligned version of new.
  // 'p' is the pointer created
  // 'page_aligned_p' is the page-aligned pointer created
  void pa_new(unsigned int num_bytes, char *&p, char *&page_aligned_p)
  {
#ifdef NON_UNIX
    SYSTEM_INFO system_info;
    GetSystemInfo(&system_info);
    long page_size = system_info.dwPageSize;
#else
    int page_size = getpagesize();
#endif
    p = ::new char [num_bytes + page_size];

    page_aligned_p = (char *)((((unsigned long)p + page_size - 1) / page_size) * page_size);
  }
};

CGlobalItems::CGlobalItems(bool *exitFlag)
 : quiet(false)
 , fast(false)
 , sync_bonnie(false)
#ifdef O_DIRECT
 , use_direct_io(false)
#endif
 , timer()
 , ram(0)
 , sem(SemKey, TestCount)
 , name(NULL)
 , bufSync(false)
 , chunk_bits(DefaultChunkBits)
 , doExit(exitFlag)
 , m_chunk_size(DefaultChunkSize)
 , m_buf(NULL)
 , m_buf_pa(NULL)
{
  pa_new(m_chunk_size, m_buf, m_buf_pa);
  SetName(".");
}

void CGlobalItems::decrement_and_wait(int nr_sem)
{
  if(sem.decrement_and_wait(nr_sem))
    exit(1);
}

int TestDirOps(int directory_size, int max_size, int min_size
             , int num_directories, CGlobalItems &globals);
int TestFileOps(int file_size, CGlobalItems &globals);

static bool exitNow;
static bool already_printed_error;

#ifdef USE_SA_SIGACTION
#define SIGNAL_NUMBER siginf->si_signo
#else
#define SIGNAL_NUMBER sig
#endif

extern "C"
{
  void ctrl_c_handler(int sig
#ifdef USE_SA_SIGACTION
		    , siginfo_t *siginf, void *unused
#endif
		     )
  {
    if(SIGNAL_NUMBER == SIGXCPU)
      fprintf(stderr, "Exceeded CPU usage.\n");
    else if(SIGNAL_NUMBER == SIGXFSZ)
      fprintf(stderr, "exceeded file storage limits.\n");
    exitNow = true;
  }
}

int main(int argc, char *argv[])
{
  int    file_size = DefaultFileSize;
  int    directory_size = DefaultDirectorySize;
  int    directory_max_size = DefaultDirectoryMaxSize;
  int    directory_min_size = DefaultDirectoryMinSize;
  int    num_bonnie_procs = 0;
  int    num_directories = 1;
  int    count = -1;
  const char * machine = NULL;
  char *userName = NULL, *groupName = NULL;
  CGlobalItems globals(&exitNow);
  bool setSize = false;

  exitNow = false;
  already_printed_error = false;

  struct sigaction sa;
#ifdef USE_SA_SIGACTION
  sa.sa_sigaction = &ctrl_c_handler;
  sa.sa_flags = SA_RESETHAND | SA_SIGINFO;
#else
  sa.sa_handler = ctrl_c_handler;
  sa.sa_flags = SA_RESETHAND;
#endif
  if(sigaction(SIGINT, &sa, NULL)
   || sigaction(SIGXCPU, &sa, NULL)
   || sigaction(SIGXFSZ, &sa, NULL))
  {
    printf("Can't handle SIGINT.\n");
    return 1;
  }
#ifdef USE_SA_SIGACTION
  sa.sa_sigaction = NULL;
#endif
  sa.sa_handler = SIG_IGN;
  if(sigaction(SIGHUP, &sa, NULL))
  {
    printf("Can't handle SIGHUP.\n");
    return 1;
  }

#ifdef _SC_PHYS_PAGES
  int page_size = sysconf(_SC_PAGESIZE);
  int num_pages = sysconf(_SC_PHYS_PAGES);
  if(page_size != -1 && num_pages != -1)
  {
    globals.ram = page_size/1024 * (num_pages/1024);
  }
#else

#ifdef AIX_MEM_SIZE
  struct CuAt *odm_obj;
  int how_many;

  odm_set_path("/etc/objrepos");
  odm_obj = getattr("sys0", "realmem", 0, &how_many);
  globals.ram = atoi(odm_obj->value) / 1024;
  odm_terminate();
  printf("Memory = %d MiB\n", globals.ram);
#endif

#endif

  int int_c;
  while(-1 != (int_c = getopt(argc, argv, "bd:fg:m:n:p:qr:s:u:x:y"
#ifdef O_DIRECT
                             "D"
#endif
                             )) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'b':
        globals.bufSync = true;
      break;
      case 'd':
        if(chdir(optarg))
        {
          fprintf(stderr, "Can't change to directory \"%s\".\n", optarg);
          usage();
        }
      break;
      case 'f':
        globals.fast = true;
      break;
      case 'm':
        machine = optarg;
      break;
      case 'n':
        sscanf(optarg, "%d:%d:%d:%d", &directory_size
                     , &directory_max_size, &directory_min_size
                     , &num_directories);
      break;
      case 'p':
        num_bonnie_procs = atoi(optarg);
                        /* Set semaphore to # of bonnie++ procs 
                           to synchronize */
      break;
      case 'q':
        globals.quiet = true;
      break;
      case 'r':
        globals.ram = atoi(optarg);
      break;
      case 's':
      {
        char *sbuf = strdup(optarg);
        char *size = strtok(sbuf, ":");
        file_size = atoi(size);
        char c = size[strlen(size) - 1];
        if(c == 'g' || c == 'G')
          file_size *= 1024;
        size = strtok(NULL, "");
        if(size)
        {
          int tmp = atoi(size);
          c = size[strlen(size) - 1];
          if(c == 'k' || c == 'K')
            tmp *= 1024;
          globals.set_chunk_size(tmp);
        }
        setSize = true;
      }
      break;
      case 'g':
        if(groupName)
          usage();
        groupName = optarg;
      break;
      case 'u':
      {
        if(userName)
          usage();
        userName = strdup(optarg);
        int i;
        for(i = 0; userName[i] && userName[i] != ':'; i++)
        {}
        if(userName[i] == ':')
        {
          if(groupName)
            usage();
          userName[i] = '\0';
          groupName = &userName[i + 1];
        }
      }
      break;
      case 'x':
        count = atoi(optarg);
      break;
      case 'y':
                        /* tell procs to synchronize via previous
                           defined semaphore */
        globals.sync_bonnie = true;
      break;
#ifdef O_DIRECT
      case 'D':
                        /* open file descriptor with direct I/O */
        globals.use_direct_io = true;
      break;
#endif
    }
  }
  if(optind < argc)
    usage();

  if(globals.ram && !setSize)
  {
    if(file_size < (globals.ram * 2))
      file_size = globals.ram * 2;
    // round up to the nearest gig
    if(file_size % 1024 > 512)
      file_size = file_size + 1024 - (file_size % 1024);
  }

  if(machine == NULL)
  {
    struct utsname utsBuf;
    if(uname(&utsBuf) != -1)
      machine = utsBuf.nodename;
  }

  if(userName || groupName)
  {
    if(bon_setugid(userName, groupName, globals.quiet))
      return 1;
    if(userName)
      free(userName);
  }
  else if(geteuid() == 0)
  {
    fprintf(stderr, "You must use the \"-u\" switch when running as root.\n");
    usage();
  }

  if(num_bonnie_procs && globals.sync_bonnie)
    usage();

  if(num_bonnie_procs)
  {
    if(num_bonnie_procs == -1)
    {
      return globals.sem.clear_sem();
    }
    else
    {
      return globals.sem.create(num_bonnie_procs);
    }
  }

  if(globals.sync_bonnie)
  {
    if(globals.sem.get_semid())
      return 1;
  }

  if(file_size < 0 || directory_size < 0 || (!file_size && !directory_size) )
    usage();
  if(directory_size > 262143)
  {
    fprintf(stderr, "Maximum directory size is 976\n");
    return 1;
  }
  if(globals.chunk_size() < 256 || globals.chunk_size() > Unit)
    usage();
  int i;
  globals.chunk_bits = 0;
  for(i = globals.chunk_size(); i > 1; i = i >> 1, globals.chunk_bits++)
  {}
  if(1 << globals.chunk_bits != globals.chunk_size())
    usage();

  if( (directory_max_size != -1 && directory_max_size != -2)
     && (directory_max_size < directory_min_size || directory_max_size < 0
     || directory_min_size < 0) )
    usage();
  /* If the storage size is too big for the maximum number of files (1000G) */
  if(file_size > IOFileSize * MaxIOFiles)
    usage();
  /* If the file size is so large and the chunk size is so small that we have
   * more than 2G of chunks */
  if(globals.chunk_bits < 20 && file_size > (1 << (31 - 20 + globals.chunk_bits)) )
    usage();
  // if doing more than one test run then we print a header before the
  // csv format output.
  if(count > 1)
  {
    globals.timer.SetType(BonTimer::csv);
    globals.timer.PrintHeader(stdout);
  }
#ifdef OS2
    ULONG myPid = 0;
    DosQuerySysInfo(QSV_FOREGROUND_PROCESS, QSV_FOREGROUND_PROCESS
                  , &myPid, sizeof(myPid));
#else
  pid_t myPid = getpid();
#endif
  srand(myPid ^ time(NULL));
  for(; count > 0 || count == -1; count--)
  {
    globals.timer.Initialize();
    int rc;
    rc = TestFileOps(file_size, globals);
    if(rc) return rc;
    rc = TestDirOps(directory_size, directory_max_size, directory_min_size
                  , num_directories, globals);
    if(rc) return rc;
    // if we are only doing one test run then print a plain-text version of
    // the results before printing a csv version.
    if(count == -1)
    {
      globals.timer.SetType(BonTimer::txt);
      rc = globals.timer.DoReport(machine, file_size, directory_size
                                , directory_max_size, directory_min_size
                                , num_directories, globals.chunk_size()
                                , globals.quiet ? stderr : stdout);
    }
    // print a csv version in every case
    globals.timer.SetType(BonTimer::csv);
    rc = globals.timer.DoReport(machine, file_size, directory_size
                              , directory_max_size, directory_min_size
                              , num_directories, globals.chunk_size(), stdout);
    if(rc) return rc;
  }
}

int
TestFileOps(int file_size, CGlobalItems &globals)
{
  if(file_size)
  {
    CFileOp file(globals.timer, file_size, globals.chunk_bits, globals.bufSync
#ifdef O_DIRECT
               , globals.use_direct_io
#endif
                );
    int    num_chunks;
    int    words;
    char  *buf = globals.buf();
    int    bufindex;
    int    i;

    if(globals.ram && file_size < globals.ram * 2)
    {
      fprintf(stderr
            , "File size should be double RAM for good results, RAM is %dM.\n"
            , globals.ram);
      return 1;
    }
    // default is we have 1M / 8K * 200 chunks = 25600
    num_chunks = Unit / globals.chunk_size() * file_size;

    int rc;
    rc = file.open(globals.name, true, true);
    if(rc)
      return rc;
    if(exitNow)
      return EXIT_CTRL_C;
    globals.timer.timestamp();

    if(!globals.fast)
    {
      globals.decrement_and_wait(Putc);
      // Fill up a file, writing it a char at a time with the stdio putc() call
      if(!globals.quiet) fprintf(stderr, "Writing with putc()...");
      for(words = 0; words < num_chunks; words++)
      {
        if(file.write_block_putc() == -1)
          return 1;
        if(exitNow)
          return EXIT_CTRL_C;
      }
      fflush(NULL);
      /*
       * note that we always close the file before measuring time, in an
       *  effort to force as much of the I/O out as we can
       */
      file.close();
      globals.timer.get_delta_t(Putc);
      if(!globals.quiet) fprintf(stderr, "done\n");
    }
    /* Write the whole file from scratch, again, with block I/O */
    if(file.reopen(true))
      return 1;
    globals.decrement_and_wait(FastWrite);
    if(!globals.quiet) fprintf(stderr, "Writing intelligently...");
    memset(buf, 0, globals.chunk_size());
    globals.timer.timestamp();
    bufindex = 0;
    // for the number of chunks of file data
    for(i = 0; i < num_chunks; i++)
    {
      if(exitNow)
        return EXIT_CTRL_C;
      // for each chunk in the Unit
      buf[bufindex]++;
      bufindex = (bufindex + 1) % globals.chunk_size();
      if(file.write_block(PVOID(buf)) == -1)
        return io_error("write(2)");
    }
    file.close();
    globals.timer.get_delta_t(FastWrite);
    if(!globals.quiet) fprintf(stderr, "done\n");


    /* Now read & rewrite it using block I/O.  Dirty one word in each block */
    if(file.reopen(false))
      return 1;
    if (file.seek(0, SEEK_SET) == -1)
    {
      if(!globals.quiet) fprintf(stderr, "error in lseek(2) before rewrite\n");
      return 1;
    }
    globals.decrement_and_wait(ReWrite);
    if(!globals.quiet) fprintf(stderr, "Rewriting...");
    globals.timer.timestamp();
    bufindex = 0;
    for(words = 0; words < num_chunks; words++)
    { // for each chunk in the file
      if (file.read_block(PVOID(buf)) == -1)
        return 1;
      bufindex = bufindex % globals.chunk_size();
      buf[bufindex]++;
      bufindex++;
      if (file.seek(-1, SEEK_CUR) == -1)
        return 1;
      if (file.write_block(PVOID(buf)) == -1)
        return io_error("re write(2)");
      if(exitNow)
        return EXIT_CTRL_C;
    }
    file.close();
    globals.timer.get_delta_t(ReWrite);
    if(!globals.quiet) fprintf(stderr, "done\n");


    if(!globals.fast)
    {
      // read them all back with getc()
      if(file.reopen(false, true))
        return 1;
      globals.decrement_and_wait(Getc);
      if(!globals.quiet) fprintf(stderr, "Reading with getc()...");
      globals.timer.timestamp();

      for(words = 0; words < num_chunks; words++)
      {
        if(file.read_block_getc(buf) == -1)
          return 1;
        if(exitNow)
          return EXIT_CTRL_C;
      }

      file.close();
      globals.timer.get_delta_t(Getc);
      if(!globals.quiet) fprintf(stderr, "done\n");
    }

    /* Now suck it in, Chunk at a time, as fast as we can */
    if(file.reopen(false))
      return 1;
    if (file.seek(0, SEEK_SET) == -1)
      return io_error("lseek before read");
    globals.decrement_and_wait(FastRead);
    if(!globals.quiet) fprintf(stderr, "Reading intelligently...");
    globals.timer.timestamp();
    for(i = 0; i < num_chunks; i++)
    { /* per block */
      if ((words = file.read_block(PVOID(buf))) == -1)
        return io_error("read(2)");
      if(exitNow)
        return EXIT_CTRL_C;
    } /* per block */
    file.close();
    globals.timer.get_delta_t(FastRead);
    if(!globals.quiet) fprintf(stderr, "done\n");

    globals.timer.timestamp();
    if(file.seek_test(globals.quiet, globals.sem))
      return 1;

    /*
     * Now test random seeks; first, set up for communicating with children.
     * The object of the game is to do "Seeks" lseek() calls as quickly
     *  as possible.  So we'll farm them out among SeekProcCount processes.
     *  We'll control them by writing 1-byte tickets down a pipe which
     *  the children all read.  We write "Seeks" bytes with val 1, whichever
     *  child happens to get them does it and the right number of seeks get
     *  done.
     * The idea is that since the write() of the tickets is probably
     *  atomic, the parent process likely won't get scheduled while the
     *  children are seeking away.  If you draw a picture of the likely
     *  timelines for three children, it seems likely that the seeks will
     *  overlap very nicely with the process scheduling with the effect
     *  that there will *always* be a seek() outstanding on the file.
     * Question: should the file be opened *before* the fork, so that
     *  all the children are lseeking on the same underlying file object?
     */
  }
  return 0;
}

int
TestDirOps(int directory_size, int max_size, int min_size
         , int num_directories, CGlobalItems &globals)
{
  COpenTest open_test(globals.chunk_size(), globals.bufSync, globals.doExit);
  if(!directory_size)
  {
    return 0;
  }
  // if directory_size (in K) * data per file*2 > (ram << 10) (IE memory /1024)
  // then the storage of file names will take more than half RAM and there
  // won't be enough RAM to have Bonnie++ paged in and to have a reasonable
  // meta-data cache.
  if(globals.ram && directory_size * MaxDataPerFile * 2 > (globals.ram << 10))
  {
    fprintf(stderr
        , "When testing %dK of files in %d MiB of RAM the system is likely to\n"
           "start paging Bonnie++ data and the test will give suspect\n"
           "results, use less files or install more RAM for this test.\n"
          , directory_size, globals.ram);
    return 1;
  }
  // Can't use more than 1G of RAM
  if(directory_size * MaxDataPerFile > (1 << 20))
  {
    fprintf(stderr, "Not enough ram to test with %dK files.\n"
                  , directory_size);
    return 1;
  }
  globals.decrement_and_wait(CreateSeq);
  if(!globals.quiet) fprintf(stderr, "Create files in sequential order...");
  if(open_test.create(globals.name, globals.timer, directory_size
                    , max_size, min_size, num_directories, false))
    return 1;
  globals.decrement_and_wait(StatSeq);
  if(!globals.quiet) fprintf(stderr, "done.\nStat files in sequential order...");
  if(open_test.stat_sequential(globals.timer))
    return 1;
  globals.decrement_and_wait(DelSeq);
  if(!globals.quiet) fprintf(stderr, "done.\nDelete files in sequential order...");
  if(open_test.delete_sequential(globals.timer))
    return 1;
  if(!globals.quiet) fprintf(stderr, "done.\n");

  globals.decrement_and_wait(CreateRand);
  if(!globals.quiet) fprintf(stderr, "Create files in random order...");
  if(open_test.create(globals.name, globals.timer, directory_size
                    , max_size, min_size, num_directories, true))
    return 1;
  globals.decrement_and_wait(StatRand);
  if(!globals.quiet) fprintf(stderr, "done.\nStat files in random order...");
  if(open_test.stat_random(globals.timer))
    return 1;
  globals.decrement_and_wait(DelRand);
  if(!globals.quiet) fprintf(stderr, "done.\nDelete files in random order...");
  if(open_test.delete_random(globals.timer))
    return 1;
  if(!globals.quiet) fprintf(stderr, "done.\n");
  return 0;
}

void
usage()
{
  fprintf(stderr,
    "usage: bonnie++ [-d scratch-dir] [-s size(MiB)[:chunk-size(b)]]\n"
    "                [-n number-to-stat[:max-size[:min-size][:num-directories]]]\n"
    "                [-m machine-name]\n"
    "                [-r ram-size-in-MiB]\n"
    "                [-x number-of-tests] [-u uid-to-use:gid-to-use] [-g gid-to-use]\n"
    "                [-q] [-f] [-b] [-p processes | -y]\n"
#ifdef O_DIRECT
    "                [-D]\n"
#endif
    "\nVersion: " BON_VERSION "\n");
  exit(1);
}

int
io_error(CPCCHAR message, bool do_exit)
{
  char buf[1024];

  if(!already_printed_error && !do_exit)
  {
    sprintf(buf, "Bonnie: drastic I/O error (%s)", message);
    perror(buf);
    already_printed_error = 1;
  }
  if(do_exit)
    exit(1);
  return(1);
}

